src/Functions/array/FunctionArrayMapped.h

```diff
@@ -144,32 +161,54 @@ class FunctionArrayMapped : public IFunction
             throw Exception(
                 ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
                 "First argument for this overload of {} must be a function with {} arguments, found {} instead",
-                getName(), nested_types.size(), arguments[0]->getName());
+                getName(),
+                nested_types.size(),
+                arguments[0]->getName());
 
         arguments[0] = std::make_shared<DataTypeFunction>(nested_types);
     }
 
     DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
     {
-        size_t min_args = Impl::needExpression() ? 2 : 1;
+        size_t min_args = (num_fixed_params + Impl::needExpression()) ? 2 : 1;
         if (arguments.size() < min_args)
-            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
-                            "Function {} needs at least {} argument, passed {}",
-                            getName(), min_args, arguments.size());
-
-        if ((arguments.size() == 1) && is_argument_type_array)
+            throw Exception(
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Function {} needs at least {} argument{}, passed {}",
+                getName(),
+                min_args,
+                (min_args > 1 ? "s" : ""),
+                arguments.size());
+
+        if ((arguments.size() == 1 + num_fixed_params) && is_argument_type_array)
         {
-            const auto * data_type = checkAndGetDataType<typename Impl::data_type>(arguments[0].type.get());
+            const auto * data_type = checkAndGetDataType<typename Impl::data_type>(arguments[num_fixed_params].type.get());
 
             if (!data_type)
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The only argument for function {} must be array. "
-                                "Found {} instead", getName(), arguments[0].type->getName());
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "The {}{}{} argument for function {} must be array. Found {} instead",
+                    num_fixed_params + 1,
+                    getOrdinalSuffix(num_fixed_params + 1),
+                    (num_fixed_params == 0 ? " and only" : ""),
+                    getName(),
+                    arguments[num_fixed_params].type->getName());
+
+            if constexpr (num_fixed_params)
+                Impl::checkArguments(
+                    std::span<const ColumnWithTypeAndName, num_fixed_params>(std::begin(arguments), num_fixed_params), getName());
```

Comment:
here and in `execute` - why do we need to use `span` in this strange and confusing way? can we just pass type to `checkArguments` and limit to `execute`?

Reply:
but fixed params always have the same type `ColumnWithTypeAndName` - why do we need to unwind arguments into tuple? we can just pass iterator on the first fixed argument, and then `checkArguments` and `execute` knows the number of such params... I don't see the point of such complications